### PR TITLE
introduces first steps for draining

### PR DESCRIPTION
**Pull request by xh3b4sd**
Towards https://github.com/giantswarm/giantswarm/issues/1918.

### CODE REVIEW

File: service/controller/v11/resource/deployment/master_deployment.go [hunk 1]
```diff
@@ -19,6 +19,7 @@ func newMasterDeployments(customObject v1alpha1.KVMConfig) ([]*extensionsv1.Depl
 
 	privileged := true
 	replicas := int32(1)
+	podDeletionGracePeriod := int64(key.PodDeletionGracePeriod.Seconds())
```
Comment (xh3b4sd) [inline service/controller/v11/resource/deployment/master_deployment.go]: We prepare the master pods for draining support. This means the grace deletion period is extended to give workloads more time in the guest clusters. Below is described what else happens here. Note that all this applies to the workers as well.

File: service/controller/v11/resource/deployment/master_deployment.go [hunk 2]
```diff
@@ -89,11 +90,15 @@ func newMasterDeployments(customObject v1alpha1.KVMConfig) ([]*extensionsv1.Depl
 				Template: apiv1.PodTemplateSpec{
 					ObjectMeta: apismetav1.ObjectMeta{
 						GenerateName: key.MasterID,
+						Finalizers: []string{
+							key.DrainingNodesFinalizer,
+						},
```
Comment (xh3b4sd) [inline service/controller/v11/resource/deployment/master_deployment.go]: We manage a specific finalizer in the kvm-operator to have control over the guest cluster nodes deletion process.

File: service/controller/v11/resource/deployment/master_deployment.go [hunk 3]
```diff
@@ -89,11 +90,15 @@ func newMasterDeployments(customObject v1alpha1.KVMConfig) ([]*extensionsv1.Depl
...
 						Labels: map[string]string{
-							"app":      key.MasterID,
-							"cluster":  key.ClusterID(customObject),
-							"customer": key.ClusterCustomer(customObject),
-							"node":     masterNode.ID,
+							"app":               key.MasterID,
+							"cluster":           key.ClusterID(customObject),
+							"customer":          key.ClusterCustomer(customObject),
+							"node":              masterNode.ID,
+							key.PodWatcherLabel: "kvm-operator",
```
Comment (xh3b4sd) [inline service/controller/v11/resource/deployment/master_deployment.go]: The label is used for the pod resource which is reconciled in its own framework. The deletion magic happens there.

File: service/controller/v11/resource/pod/delete.go [hunk 4]
```diff
@@ -2,8 +2,116 @@ package pod
...
+
+	r.logger.LogCtx(ctx, "debug", "found the current version of the reconciled pod in the Kubernetes API")
+
+	// TODO drain guest cluster node and only remove the finalizer as soon as the
+	// guest cluster node associated with the reconciled host cluster pod is
+	// drained.
+
+	// TODO go ahead and do not block once the draining is initialized. Using
+	// finalizers on the pods the delete event will be replayed and we can check
+	// if the draining completed on the next reconciliation loop.
```
Comment (xh3b4sd) [inline service/controller/v11/resource/pod/delete.go]: All this is tested so far and works. The TODOs are going to be addressed in the next steps. Should be fairly simple to get the draining going via the node-operator.

**Review by xh3b4sd — COMMENTED** [review]
This is successfully tested on `geckon`. Scaling down and deletion works as expected. The finalizers are managed by the pod resource implementation.

**Review by rossf7 — APPROVED** [review]
Not deep into how this works on KVM but this LGTM.

**Pull request merged**